### PR TITLE
feat(storage): support new lru-cache so that we can control memory usage precisely

### PR DESCRIPTION
**Pull request by Little-Wallace**
## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- port a new block cache. compare moka::Cache, it will count reference for every lookup request, and until the iterator holding the block is dropped, the block cache will always keep it in LRU list and count the data in memory usage.

## TODO

* We need to add a request queue so that compute-node won't send multiple request for the same object.
* We need to change the interface and iterator can not own `Arc<Block>` by itself. It must keep `CachableEntry` during its lifetime to ensure the entry will never release by `LRUCache`.

## Checklist

- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
part of https://github.com/singularity-data/risingwave/issues/1773

### CODE REVIEW

**Comment by Little-Wallace** [comment]
run `cargo bench -- bench_lru_cache` in `src/storage` directory, and you will get the benchmark result.

```
100000 keys cost: 58.171416ms
100000 keys cost: 66.689041ms
100000 keys cost: 67.786375ms
100000 keys cost: 68.181791ms
100000 keys cost: 68.90775ms
100000 keys cost: 69.716791ms
100000 keys cost: 71.040208ms
100000 keys cost: 71.09475ms

100000 keys cost: 21.712458ms
100000 keys cost: 22.623875ms
100000 keys cost: 22.642416ms
100000 keys cost: 22.721125ms
100000 keys cost: 22.740208ms
100000 keys cost: 22.744083ms
100000 keys cost: 22.789666ms
100000 keys cost: 22.837916ms
```
The data above  is `moka::Cache` while the data below is current pr.

File: src/storage/src/hummock/cache.rs [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+
+use std::ptr::null_mut;
+use std::sync::Arc;
+
+use spin::Mutex;
+
+const IN_CACHE: u8 = 1;
+const REVERSE_IN_CACHE: u8 = !IN_CACHE;
+
+pub struct LRUHandle<K: PartialEq + Default, T> {
```
Comment (BugenZhao) [inline src/storage/src/hummock/cache.rs:23]: https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case :)
```suggestion
pub struct LruHandle<K: PartialEq + Default, T> {
```

Comment (Little-Wallace) [reply src/storage/src/hummock/cache.rs]: Done

**Review by hzxa21 — COMMENTED** [review]
Looked at `LruHandle` & `LRUCacheShard` and left some comments. Generally LGTM. Will finish reviewing the remaining parts tmr.

File: src/storage/src/hummock/cache.rs [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+            null_mut()
+        } else {
+            let ptr = Box::into_raw(handle);
+            let old = self.table.insert(hash, ptr);
+            if !old.is_null() {
+                let data = self.remove_cache_handle(old).unwrap();
+                last_reference_list.push(data);
+            }
+            self.usage += charge;
+            (*ptr).add_ref();
```
Comment (hzxa21) [inline src/storage/src/hummock/cache.rs:300]: We only `add_ref()` once when the handle is inserted but not `add_ref()` for the returned handle in L413. Will this be a problem?

File: src/storage/src/hummock/cache.rs (hunk 1 above)
Comment (BugenZhao) [reply src/storage/src/hummock/cache.rs]: Could you please fix other occurrences of `LRU` in this PR?

File: src/storage/src/hummock/cache.rs [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+        let h = self.table.remove(hash, key);
+        if !h.is_null() {
+            (*h).set_in_cache(false);
+            if !(*h).has_refs() {
+                self.lru_remove(h);
+                let mut node = Box::from_raw(h);
+                node.set_in_cache(false);
+                self.usage -= node.charge;
+                let data = node.value.take();
+                self.recyle_handle_object(node);
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/cache.rs]: done

File: src/storage/src/hummock/cache.rs [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+        assert_eq!(count, self.elems);
+        self.list = new_list;
+    }
+}
+
+pub struct LRUCacheShard<K: PartialEq + Default, T> {
+    lru: Box<LruHandle<K, T>>,
+    table: LruHandleTable<K, T>,
+    object_pool: Vec<Box<LruHandle<K, T>>>,
+    lru_usage: usize,
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/cache.rs:194]: yes. We can get pinned memory(the total memory of blocks held by iterator) by `usage - lru_usage` in the future.

File: src/storage/src/hummock/cache.rs (hunk 2 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/cache.rs:300]: Emmm, why? We only need `add_ref()` once.

File: src/storage/src/hummock/cache.rs [hunk 5]
```diff
@@ -0,0 +1,664 @@
...
+    unsafe fn insert(&mut self, hash: u64, h: *mut LruHandle<K, T>) -> *mut LruHandle<K, T> {
+        let idx = (hash as usize) & (self.list.len() - 1);
+        let (mut prev, ptr) = self.find_pointer(idx, &(*h).key);
+        if prev.is_null() {
+            let idx = (hash as usize) & (self.list.len() - 1);
+            self.list[idx] = h;
+        } else {
+            (*prev).next_hash = h;
+        }
+        if !ptr.is_null() && (*ptr).key.eq(&(*h).key) {
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/cache.rs]: yes.

File: src/storage/src/hummock/cache.rs [hunk 6]
```diff
@@ -0,0 +1,664 @@
...
+        }
+        self.elems -= 1;
+        ptr
+    }
+
+    unsafe fn insert(&mut self, hash: u64, h: *mut LruHandle<K, T>) -> *mut LruHandle<K, T> {
+        let idx = (hash as usize) & (self.list.len() - 1);
+        let (mut prev, ptr) = self.find_pointer(idx, &(*h).key);
+        if prev.is_null() {
+            let idx = (hash as usize) & (self.list.len() - 1);
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/cache.rs]: done

**Review by hzxa21 — APPROVED** [review]

File: src/storage/Cargo.toml [hunk 7]
```diff
@@ -49,6 +49,7 @@ risingwave_rpc_client = { path = "../rpc_client" }
 # ], optional = true }
 serde = { version = "1", features = ["derive"] }
 smallvec = "1"
+spin = "0.5.2"
```
Comment (MrCroxx) [inline src/storage/Cargo.toml:52]: spin 0.5.2 seems too old (current version = 0.9.3), can its version be updated?

**Comment by MrCroxx** [comment]
BTW, a random question, is it fine to use lots of CAS operations with tokio? IMO, CAS will not suspend the fiber, and if conflict happens frequently, some fiber cannot be scheduled, 🤔.

File: src/storage/Cargo.toml (hunk 7 above)
Comment (Little-Wallace) [reply src/storage/Cargo.toml:52]: Emmmm, I just changed it because other library depend on 0.5.2

**Comment by skyzh** [comment]
Would you please also run the benchmark with thread / leak / address sanitizer, so that we can probably find some unsound issue before merging?

File: src/storage/src/hummock/cache.rs [hunk 8]
```diff
@@ -0,0 +1,664 @@
...
+
+use std::ptr::null_mut;
+use std::sync::Arc;
+
+use spin::Mutex;
+
+const IN_CACHE: u8 = 1;
+const REVERSE_IN_CACHE: u8 = !IN_CACHE;
+
+pub struct LruHandle<K: PartialEq + Default, T> {
```
Comment (wenym1) [inline src/storage/src/hummock/cache.rs:23]: Can we have some doc for the struct in this doc? Like the overall idea, code structure and the explanation for struct fields and method parameters.

File: src/storage/src/hummock/cache.rs [hunk 9]
```diff
@@ -0,0 +1,692 @@
...
+// LruCacheShard::Release on entry in state 1, it will go into state 2.
+// To move from state 1 to state 3, either call LruCacheShard::Erase or
+// LruCacheShard::Insert with the same key (but possibly different value).
+// To move from state 2 to state 1, use LruCacheShard::Lookup.
+// Before destruction, make sure that no handles are in state 1. This means
+// that any successful LruCacheShard::Lookup/LruCacheShard::Insert have a
+// matching LruCache::Release (to move into state 2) or LruCacheShard::Erase
+// (to move into state 3).
+pub struct LruHandle<K: PartialEq + Default, T> {
+    next_hash: *mut LruHandle<K, T>,
```
Comment (skyzh) [inline src/storage/src/hummock/cache.rs:52]: better to doc the fields.

next_hash: next element in the hash bucket the LRU handle represents
next: next element in LRU linked list
prev: prev element in LRU linked list
etc.

File: src/storage/src/hummock/cache.rs [hunk 10]
```diff
@@ -0,0 +1,692 @@
...
+                false,
+            )));
+        }
+        Self {
+            num_shard_bits,
+            shards,
+        }
+    }
+
+    pub fn lookup(self: &Arc<Self>, hash: u64, key: &K) -> Option<CachableEntry<K, T>> {
```
Comment (skyzh) [inline src/storage/src/hummock/cache.rs:417]: Why lookup needs to provide a hash? Users are expected to calculate hash by themselves?

File: src/storage/src/hummock/cache.rs [hunk 11]
```diff
@@ -0,0 +1,692 @@
...
+
+    pub fn erase(&self, hash: u64, key: &K) {
+        let data = unsafe {
+            let mut shard = self.shards[self.shard(hash)].lock();
+            shard.erase(hash, key)
+        };
+        drop(data);
+    }
+
+    pub fn get_memory_usage(&self) -> usize {
```
Comment (skyzh) [inline src/storage/src/hummock/cache.rs:473]: /// `get_memory_usage` will hold the lock for every bucket.

File: src/storage/src/hummock/cache.rs [hunk 12]
```diff
@@ -0,0 +1,692 @@
...
+// that any successful LruCacheShard::Lookup/LruCacheShard::Insert have a
+// matching LruCache::Release (to move into state 2) or LruCacheShard::Erase
+// (to move into state 3).
+pub struct LruHandle<K: PartialEq + Default, T> {
+    next_hash: *mut LruHandle<K, T>,
+    next: *mut LruHandle<K, T>,
+    prev: *mut LruHandle<K, T>,
+    key: K,
+    hash: u64,
+    value: Option<T>,
```
Comment (skyzh) [inline src/storage/src/hummock/cache.rs:57]: why value needs to be Option?

**Review by skyzh — COMMENTED** [review]
LGTM for the rest.

**Comment by skyzh** [comment]
Better to verify test cases and benches with thread / leak / address sanitizer. Also cargo-miri might be helpful in case of pointer operations.

**Comment by Little-Wallace** [comment]
I run ` DYLD_INSERT_LIBRARIES=/Users/liuwei/.rustup/toolchains/nightly-2022-04-09-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/librustc-nightly_rt.asan.dylib cargo bench -Zbuild-std --target aarch64-apple-darwin  ` successfully in src/storage and get nothing error

File: src/storage/src/hummock/cache.rs (hunk 12 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/cache.rs:57]: Because I need a head pointer....

File: src/storage/src/hummock/cache.rs (hunk 10 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/cache.rs:417]: I will use a BlockCache to wrapper  Cache<(u64,u64), Box<Block>> and provide a hash algorithm there.

File: src/storage/src/hummock/cache.rs (hunk 12 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/cache.rs:57]: And I could take it from `LruHandle` and then destruct them outside mutex.

File: src/storage/src/hummock/cache.rs [hunk 13]
```diff
@@ -0,0 +1,692 @@
...
+            hash: 0,
+            value: None,
+            charge: 0,
+            refs: 0,
+            flags: 0,
+        }
+    }
+}
+
+impl<K: PartialEq + Default, T> LruHandle<K, T> {
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/cache.rs:79]: Emmmm. Will you give a K::Sync as key of BTreeMap ?

[the remark this answers is not in the document]
Comment (skyzh) [reply src/storage/src/hummock/cache.rs:85]: You're right 🤣

**Review by skyzh — APPROVED** [review]

File: src/storage/src/hummock/cache.rs (hunk 1 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/cache.rs]: done

**Pull request merged**